feat: Import all DocTypes from single namespace

gavindsouza opened the pull request:
### Current Implementation

The `frappe.doctypes` namespace will act as a proxy for accessing all controllers for the DocTypes you'd want to import. This aims to get rid of the painful imports we have to do using something like `from frappe.desk.doctype.note.note import Note`.

There are two ways of doing this at current state of the implementation

<img width="720" alt="Screenshot 2021-08-31 at 8 37 20 PM" src="https://user-images.githubusercontent.com/36654812/131528819-9eb5c1e9-0785-4318-93ed-22c28e55740b.png">

#### Why/When you'd prefer using the full path imports instead of this?

We haven't figured out how we could make providing type hints available through imports like this at this point. So, since this is trying to aid development to a certain degree, it's disabling these auto-suggestions like _"hey, are you trying to make to call to `ToDo.destroy_this_server` or `ToDo.delete`?"_ Maybe auto-pilot can handle it even with this but I didn't test it with it LOL.

### Why this?

Primarily, because I hate remembering or going through the repo tree to figure out the namespaces to import a particular thing. This happens while writing tests or writing DocType controllers that inherit from another, in core or "custom" Frappe applications. This is an attempt only to simplify that.

<details><summary><b>Initial Draft</b></summary>
<p>

When server starts, load and cache DocType -> Controller map. Load it into Python process when imports made from `frappe.doctypes`

Eg:

```python
In [1]: from frappe.doctypes import Note

In [2]: Note
Out[2]: frappe.desk.doctype.note.note.Note

In [3]: Note.__dict__
Out[3]:
mappingproxy({'__module__': 'frappe.desk.doctype.note.note',
              'autoname': <function frappe.desk.doctype.note.note.Note.autoname(self)>,
              'validate': <function frappe.desk.doctype.note.note.Note.validate(self)>,
              'before_print': <function frappe.desk.doctype.note.note.Note.before_print(self, settings=None)>,
              '__doc__': None})
```

</p>
</details>

gavindsouza committed:
feat: Import all Frappe DocTypes from single namespace

When server starts, load and cache DocType -> Controller map. Load
it into Python process when imports made from `frappe.doctypes`

Eg:

In [1]: from frappe.doctypes import Note

In [2]: Note
Out[2]: frappe.desk.doctype.note.note.Note

In [3]: Note.__dict__
Out[3]:
mappingproxy({'__module__': 'frappe.desk.doctype.note.note',
              'autoname': <function frappe.desk.doctype.note.note.Note.autoname(self)>,
              'validate': <function frappe.desk.doctype.note.note.Note.validate(self)>,
              'before_print': <function frappe.desk.doctype.note.note.Note.before_print(self, settings=None)>,
              '__doc__': None})

gavindsouza committed:
fix: Undo module popping if no controllers found

gavindsouza committed:
Merge branch 'develop' into from-frappe-doctypes

ankush commented on frappe/doctypes.py:
```diff
@@ -0,0 +1,53 @@
+	"""
+	return dt.replace(" ", "").replace("-", "_")
+
+
+def generate_doctype_map() -> Dict[str, Document]:
```
```suggestion

def generate_doctype_map() -> Dict[str, Type[Document]]:
```
The values are classes, not instances of those classes so it should be `Type` of `Document` (or its subtypes).

---
In general, what is the usecase/benefit of doing this? 🤔

gavindsouza replied:
Updated the description about the why

gavindsouza committed:
refactor(typing): Update return type hint

The values are classes, not instances of those classes so it should be Type of Document (or its subtypes)

Co-authored-by: Ankush <ankushmenat@gmail.com>

gavindsouza committed:
chore: Add missing import

szufisher commented on frappe/doctypes.py:
```diff
@@ -0,0 +1,53 @@
+
+# loads controllers from Redis and save it under the frappe.doctypes namespace
+for dt, ct in BENCH_CONTROLLERS.items():
+	globals()[modulize(dt)] = ct
```
Controller class already available in globals(), seems it can be directly accessed like below
(no need to import from frappe.doctypes)
e.g
 `User('User','Administrator')`

Also please consider adapt the Document __init__ method to set self.doctype= type(self).__name__ if args length is 1 and type is string
```
class Document(BaseDocument):
	"""All controllers inherit from `Document`."""
	def __init__(self, *args, **kwargs):
		"""Constructor.

		:param arg1: DocType name as string or document **dict**
		:param arg2: Document name, if `arg1` is DocType name.

		If DocType name and document name are passed, the object will load
		all values (including child documents) from the database.
		"""
		self.doctype = self.name = None
		self._default_new_docs = {}
		self.flags = frappe._dict()

		if args and args[0] and isinstance(args[0], string_types):
			# first arugment is doctype
			if len(args)==1:
				# single
				self.doctype = self.name = args[0]
```
then finally pass only docname to Controller class to get the doc. 
`User('Administrator')`

ankush replied:
> Also please consider adapt the Document init method to set self.doctype= type(self).name if args length is 1 and type is string

check https://github.com/frappe/frappe/pull/13223

leela commented:
@gavindsouza , I feel using globals and caching is little heavy for this task when all we need is calling get_controller function. what about a class with __getattr__ dunder method that returns the module by calling get_controller function.

gavindsouza commented:
> @gavindsouza , I feel using globals and caching is little heavy for this task when all we need is calling get_controller function. what about a class with **getattr** dunder method that returns the module by calling get_controller function.

@leela A bit unclear about how that'd be. Can you share some code perhaps?

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into from-frappe-doctypes

gavindsouza commented:
@frappe/frappe-review-team Any ideas/concerns on this?

leela commented:
> > @gavindsouza , I feel using globals and caching is little heavy for this task when all we need is calling get_controller function. what about a class with **getattr** dunder method that returns the module by calling get_controller function.
> 
> @leela A bit unclear about how that'd be. Can you share some code perhaps?

Hi Gavin, here is what I meant

frappe/__init__.py
```
class XXX:
     def __getattr__(self, name):
         from frappe.model.base_document import get_controller
         return get_controller(name)

doctypes = XXX()
```

Here is how import works:
```
>>> from frappe import doctypes
>>>> doctypes.ToDo
 frappe.desk.doctype.todo.todo.ToDo
```

gavindsouza commented:
@leela oh, thanks. Yeah, I'd thought of this before but the API didn't feel aesthetic enough. So, looked into using custom import loaders, which didn't lead anywhere fruitful and then finally reached this PR.

I prefer `from frappe.doctypes import Note` over having to use `frappe.doctypes.Note`. But I'm not happy with the current implementation either _(obvious as this PR has been in this same state for 2 months now)_. If the former isn't possible without all this extra baggage, then I'll have to settle on the latter.

gavindsouza committed:
Merge branch 'develop' into from-frappe-doctypes

gavindsouza committed:
chore: Drop doctypes module in favour of class approach

gavindsouza committed:
Merge branch 'develop' into from-frappe-doctypes

gavindsouza committed:
feat: frappe.doctypes namespace for all doctypes

gavindsouza committed:
chore: Update frappe module str

gavindsouza committed:
refactor: frappe.doctypes namespaces

* feat: Use doctype names to fetch controllers using
  frappe.doctypes.get("DocType Name")
* You can use the controller name directly as an attribute of
  frappe.doctypes like frappe.doctypes.DocTypeName

gavindsouza committed:
refactor(minor): Use frappe.doctypes.* instead of full imports

gavindsouza committed:
style: Trim the fat (trailing white-spaces)

ankush commented:
This collapses all namespaces into one and that too dynamically with import M A G I C. 

IMHO this shouldn't be done. 

> Namespaces are one honking great idea -- let's do more of those!

How much time do we actually spend writing import statements? Even if it's long you can either set up auto-import in your editor or just spend 10 seconds looking up the import path. 

Losing the ability to use all the tools which rely on static imports for just minor one-time convenience doesn't feel justified to me. While reading the code I want to jump to from where it was imported, will my editor be able to do that after this change? Most likely not without tweaking IntelliSense a lot. We spend more time reading code than writing, so this is actually less productive in that regard. 


---

Aside: we could change the way we are structuring our code so the module name doesn't repeat twice (once in folder name and then again in filename) 

Another option is just to import them in `module/__init__.py` MANUALLY (and via boilerplate generation logic in future) That way you can do `from erpnext.stock import Item`. 

I am against this also. All of these are solutions to a problem that doesn't exist.

gavindsouza commented:
Having this common namespace was an old-er idea based on all the annoyances I faced while writing stuff like 

* `from guinea.compressor.doctype.compressor_settings.test_compressor_settings import create_test_compressor_settings`
* `from erpnext_support.erpnext_support.doctype.team_member_impersonation.team_member_impersonation import tester_bot`

This was before pylance started suggesting it'd import any class from anywhere I want. I'm not comfortable with this current implementation of this, I'm not sure if it's the whole idea either. Closing this because I don't have the same motivation to do this now that I haven't had to write really long import statements for a while.
